### PR TITLE
[layout] Insure that hyperlinks within HTML-enabled labels are exported as such within PDFs

### PR DESCRIPTION
**Pull request by nirvn**
## Description

This PR implements the agreed upon path forward (see https://github.com/qgis/QGIS/issues/39660) to unlock functional hyperlinks when exporting layouts to PDFs.

The approach is as follow: as of this PR, when creating *new* label items, the HTML rendering mode will rely on QTextDocument (instead of QtWebKit). That gives us functional hyperlinks, at the cost of a less modern HTML rendering engine. However, since we have HTML items, we really don't lose anything.

To insure that preexisting layout labels within projects are not regressing, QGIS will continue rendering those with the webkit engine. To do so, we automatically migrate old HTML-enabled label items into html items. 

_Funded by Kanton Schaffhausen._

### CODE REVIEW

File: src/core/layout/qgslayoutitemlabel.cpp [hunk 1]
```diff
@@ -123,6 +124,11 @@ void QgsLayoutItemLabel::draw( QgsLayoutItemRenderContext &context )
   {
     rectScale = context.renderContext().scaleFactor();
   }
+  else if ( mMode == QgsLayoutItemLabel::ModeHtml && mHtmlUsesTextDocument )
+  {
+    rectScale = context.renderContext().scaleFactor() * 3.67;
+    painter->scale( context.renderContext().scaleFactor() / 3.67, context.renderContext().scaleFactor() / 3.67 );
```
Comment (nyalldawson) [inline src/core/layout/qgslayoutitemlabel.cpp:130]: Where's the 3.67 value come from?

Comment (nirvn) [reply src/core/layout/qgslayoutitemlabel.cpp:105]: @nyalldawson , yours truly has no answer here; the value has been identified through trial and error, iterating until a perfect match between that and the non-HTML rendering was achieved. 

I suppose it's some unit conversion between QgsTextDocument and our rendering context. I've made a note about the mysterious (and valid) nature of the value.

File: src/core/layout/qgslayout.cpp [hunk 2]
```diff
@@ -1106,6 +1108,21 @@ QList< QgsLayoutItem * > QgsLayout::addItemsFromXml( const QDomElement &parentEl
       }
     }
 
+    if ( QgsProjectVersion( 3, 32, 0 ) > mProject->lastSaveVersion() )
+    {
+      QgsLayoutItemLabel *label = dynamic_cast<QgsLayoutItemLabel *>( item.get() );
```
Comment (m-kuhn) [inline src/core/layout/qgslayout.cpp:1113]:
```suggestion
      QgsLayoutItemLabel *label = qobject_cast<QgsLayoutItemLabel *>( item.get() );
```

File: src/core/layout/qgslayoutitemlabel.cpp [hunk 3]
```diff
@@ -39,10 +39,11 @@
 #include <QCoreApplication>
 #include <QDate>
 #include <QDomElement>
-#include <QPainter>
-#include <QTimer>
 #include <QEventLoop>
```
Comment (m-kuhn) [inline src/core/layout/qgslayoutitemlabel.cpp:42]: I think that can go too

File: src/core/layout/qgslayoutitemlabel.cpp [hunk 4]
```diff
@@ -39,10 +39,11 @@
...
+#include <QPainter>
+#include <QTextDocument>
 #include <QThread>
```
Comment (m-kuhn) [inline src/core/layout/qgslayoutitemlabel.cpp:45]: And that?

File: src/core/layout/qgslayoutitemlabel.cpp [hunk 5]
```diff
@@ -39,10 +39,11 @@
...
+#include <QTimer>
```
Comment (m-kuhn) [inline src/core/layout/qgslayoutitemlabel.cpp:46]: Also leftover?

**Review by m-kuhn — APPROVED** [review]

File: src/core/layout/qgslayoutitemlabel.cpp [hunk 6]
```diff
@@ -118,42 +93,49 @@ void QgsLayoutItemLabel::draw( QgsLayoutItemRenderContext &context )
...
+                          ( yPenAdjust + mMarginY ) * rectScale,
+                          ( rect().width() - 2 * xPenAdjust - 2 * mMarginX ) * rectScale,
+                          ( rect().height() - 2 * yPenAdjust - 2 * mMarginY ) * rectScale );
   }
   else
   {
-    // painter is scaled to dots, so scale back to layout units
-    painter->scale( context.renderContext().scaleFactor(), context.renderContext().scaleFactor() );
+    // The 3.77 adjustment value was found through trial and error, the author has however no clue as to where it comes from
+    const double adjustementFactor = 3.77;
```
Comment (nyalldawson) [inline src/core/layout/qgslayoutitemlabel.cpp:112]:
```suggestion
    const double adjustmentFactor = 3.77;
```

**Comment by nyalldawson** [comment]
What's happening with the change of font in expected_importComposerTemplate_0.png?

**Comment by nirvn** [comment]
@m-kuhn , @nyalldawson , thanks, extra comments all addressed.

@nyalldawson , regarding the chance in importComposerTemplate, it has to do with webkit-based HTML rendering not supporting negative margins, whereas the new QTextDocument rendering does (due to the fact that it doesn't respect stylesheet margin-top values, we are setting the vertical margin via painter translate, which allows us to have negative margins). Ironically this gives us an imported item that _better_ matches the original composer template :)

**Comment by agiudiceandrea** [comment]
hI @nirvn, it seems to me the hyperlinks (`<a href="url">link</a>` ) in a Label item (with Render as HTML option set) don't work when exporting the layout to PDF as GeoPDF. Is it intended?

**Comment by m-kuhn** [comment]
IIRC GeoPDF are generated through GDAL (see https://changelog.qgis.org/en/entry/1277), so that takes a different code path (CC @rouault )

**Comment by agiudiceandrea** [comment]
@nirvn, may you please have a look at the issue report https://github.com/qgis/QGIS/issues/62796?

**Pull request merged**